pkg/systemstatsmonitor/osfeature_collector.go

```diff
@@ -0,0 +1,108 @@
+	if osFeatureConfig.MetricConfigs["system/os_feature"].DisplayName != "" {
+		oc.osFeature, err = metrics.NewInt64Metric(
+			metrics.OsFeatureID,
+			osFeatureConfig.MetricConfigs[string(metrics.OsFeatureID)].DisplayName,
+			"Guest OS Features",
+			"1",
+			metrics.LastValue,
+			[]string{featureLabel, "value"})
+		if err != nil {
+			glog.Fatalf("Error initializing metric for system/os_feature: %v", err)
+		}
+	}
+	return &oc
+}
+
+func (os *osFeatureCollector) recordFeaturesFromCmdline(cmdlineArgs []system.CmdlineArg) {
+	var featuresMap = map[string]int64{
+		"KTD":                    0,
+		"UnifiedCgroupHierarchy": 0,
+		"ModuleSigned":           0,
+		"LoadPinEnabled":         0,
+	}
+	for _, cmdlineArg := range cmdlineArgs {
+		if cmdlineArg.Key == "csm.enabled" {
+			featuresMap["KTD"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+		if cmdlineArg.Key == "systemd.unified_cgroup_hierarchy" {
+			featuresMap["UnifiedCgroupHierarchy"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+		if cmdlineArg.Key == "module.sig_enforce" {
+			featuresMap["ModuleSigned"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+		if cmdlineArg.Key == "loadpin.enabled" {
+			featuresMap["LoadPinEnabled"], _ = strconv.ParseInt(cmdlineArg.Value, 10, 64)
+		}
+	}
+	// Record the values
+	os.osFeature.Record(map[string]string{featureLabel: "KTD"}, featuresMap["KTD"])
+	os.osFeature.Record(map[string]string{featureLabel: "UnifiedCgroupHierarchy"}, featuresMap["UnifiedCgroupHierarchy"])
+	if featuresMap["ModuleSigned"] == 1 && featuresMap["LoadPinEnabled"] == 1 {
+		os.osFeature.Record(map[string]string{featureLabel: "KernelModuleIntegrity"}, 1)
+	} else {
+		os.osFeature.Record(map[string]string{featureLabel: "KernelModuleIntegrity"}, 0)
+	}
+}
+
+func (os *osFeatureCollector) recordFeaturesFromModules(modules []system.Module) {
+	defaultModules, err := util.GetDefaultModules()
+	if err != nil {
+		glog.Warningf("Error retrieving default modules %v", err)
+	}
+	unknownModules := []string{}
+	for _, module := range modules {
+		if module.OutOfTree || module.Proprietary || module.Unsigned {
+			if !find(module.ModuleName, defaultModules) {
+				unknownModules = append(unknownModules, module.ModuleName)
+			}
+		}
+	}
+	if len(unknownModules) > 0 {
+		os.osFeature.Record(map[string]string{featureLabel: "UnknownModules", valueLabel: strings.Join(unknownModules, ",")}, 1)
+	} else {
+		os.osFeature.Record(map[string]string{featureLabel: "UnknownModules"}, 0)
+	}
+}
+
+func (os *osFeatureCollector) collect() {
+	cmdlineArgs, err := system.CmdlineArgs()
+	if err != nil {
+		glog.Fatalf("Error retrieving cmdline args: %v", err)
+	}
+	os.recordFeaturesFromCmdline(cmdlineArgs)
+	modules, err := system.Modules()
+	if err != nil {
+		glog.Fatalf("Error retrieving kernel modules: %v", err)
+	}
+	os.recordFeaturesFromModules(modules)
+}
+
+func find(key string, values []string) bool {
```

Comment:
This can be a public method in some `util` package

Reply:
added in a common class.